src/dd/StateGeneration.cpp

```diff
@@ -0,0 +1,145 @@
+                 Package& dd) {
+  constexpr double eps = std::numeric_limits<double>::epsilon();
+
+  const auto alpha = randomComplexOnUnitCircle(gen, dist);
+  const auto beta = randomComplexOnUnitCircle(gen, dist);
+  const auto norm = std::sqrt(2);
+
+  const std::array<vEdge, RADIX> edges{
+      vEdge::terminal(dd.cn.lookup(alpha / norm)),
+      vEdge::terminal(dd.cn.lookup(beta / norm)),
+  };
+
+  // Check: Properly normalized.
+  assert(std::sqrt(ComplexNumbers::mag2(edges[0].w) +
+                   ComplexNumbers::mag2(edges[1].w)) < 1 + eps);
+
+  return dd.makeDDNode(v, edges);
+}
+} // namespace
+
+VectorDD generateExponentialState(const std::size_t levels, Package& dd) {
+  std::random_device rd;
+  return generateExponentialState(levels, rd(), dd);
+}
+
+VectorDD generateExponentialState(const std::size_t levels,
+                                  const std::size_t seed, Package& dd) {
+  std::vector<std::size_t> nodesPerLevel(levels - 1); // [2, 4, 8, ...]
+  std::generate(nodesPerLevel.begin(), nodesPerLevel.end(),
+                [exp = 1]() mutable { return std::pow(2, exp++); });
+  return generateRandomState(levels, nodesPerLevel, ROUNDROBIN, seed, dd);
+}
+
+VectorDD generateRandomState(const std::size_t levels,
+                             const std::vector<std::size_t>& nodesPerLevel,
+                             const GenerationWireStrategy strategy,
+                             Package& dd) {
+  std::random_device rd;
+  return generateRandomState(levels, nodesPerLevel, strategy, rd(), dd);
+}
+
+VectorDD generateRandomState(const std::size_t levels,
+                             const std::vector<std::size_t>& nodesPerLevel,
+                             const GenerationWireStrategy strategy,
+                             const std::size_t seed, Package& dd) {
+  assert(levels > 0U && "Number of levels must be greater than zero");
+  assert(nodesPerLevel.size() == levels - 1 &&
+         "Number of levels - 1 must match nodesPerLevel size");
+
+  Generator gen(seed);
+  AngleDistribution dist{0, 2. * qc::PI};
+
+  auto v = static_cast<Qubit>(levels - 1);
+  const vEdge root = randomNode(v, gen, dist, dd);
+
+  std::vector<vEdge> prev{root};
+  for (const std::size_t n : nodesPerLevel) {
+    --v;
+
+    std::vector<vEdge> curr(n); // Random nodes on layer v.
+    std::generate(curr.begin(), curr.end(),
+                  [&] { return randomNode(v, gen, dist, dd); });
+
+    std::vector<std::size_t> indices(2 * prev.size()); // Indices for wireing.
+    switch (strategy) {
+    case ROUNDROBIN: {
+      std::generate(indices.begin(), indices.end(),
+                    [&n, r = std::size_t{0}]() mutable { return (r++) % n; });
+      break;
+    }
+    case RANDOM: {
+      IndexDistribution indexDist{0, curr.size() - 1};
+      std::generate(indices.begin(), indices.end(),
+                    [&indexDist, &gen]() { return indexDist(gen); });
+      break;
+    }
+    }
```

Comment:
I think this is missing one important part of the original branch, but I might be wrong about that: it needs to be ensured that each node is actually somewhat connected to the root, meaning that there are no hanging nodes which are entirely disconnected.

The round-robin one ensures that (I think). The random one does not. It could very well end up not selecting a particular node at all.

Reply:
I was wondering about this. Shouldn't a call to `dd.garbageCollect()` remove these hanging nodes (since they have a ref count of 0)?

Reply:
Nevertheless, I just figured, if one specifies a particular level to have 3 nodes and the resulting DD has only 2 - that would be quite surprising and unintuitive. I will look into this.